### DISCUSSION TITLE
How can you prevent bootstrap version collisions using shadow dom

### DISCUSSION THREAD

**Discussion by bubblypatrick** [twbs/bootstrap#36777] [Q&A]
Hello everyone hope you're all having a good day.

I'm making a web component that uses bootstrap 5. Since the shadow DOM is supposed to encapsulate styling I assumed that it would be a good way to prevent it from colliding with any sites that might be using older versions of bootstrap (since they tend to have the same class name). I feel like I'm missing something obvious, but whenever I include other versions of bootstrap in the html document they override what's in my component. I've been stuck on it for a few days now and haven't found anything to explain why this might be.

```const template = document.createElement('template');
template.innerHtml = `
  <style>@import "https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/css/bootstrap.min.css" </style>
  <div class="container">
    <p>All my web component stylings go on from here and use bootstrap classes as well</p>
  </div>
`

class MyWebComponent extends HTMLElement {
  constructor() {
    super();
    var shadow = this.attachShadow({ mode: 'open' });
    shadow.appendChild(template.content.cloneNode(true));
  }
}
window.customElements.define('my-web-component', MyWebComponent);
```

And my index.html file is simply
```
<head>
  <link rel="stylesheet" href="link to bootstrap3 or boostrap4.css" />
  <script type="module" src="linkToMyWebComponent"></script>
</head>
<body>
  <my-web-component />
</body>
```

I've noticed that if I manually define styles within the template html like `<style>.form-control { p: 10 }</style>`  that it works just fine and overwrites all the stylings. Any help, or learning material would be much appreciated. Although feel like I've read at least two dozen long articles about templates, shadow dom, etc. on it at this point. The flattened shadow dom looks like it should be encapsulated from the older scripts, but no luck.

Thank you in advance!

**Comment by TravelTimN** [comment]
@bubblypatrick 
Not entirely certain here, but something to consider is the fact that Bootstrap version 5 completely changes syntax from previous versions.

The newer version (5+) adds `-bs-` into the component data attributes, and it will not work with older versions. I see this daily with some junior developers that I mentor... Incorporating older CDN links, but grabbing version 5 components.

Perhaps food for thought on your issue, could have something to do with it...?

Example:

Version 3.4.x Modal
`data-toggle="modal"`

Version 4.6.x Modal
`data-toggle="modal"`

Version 5.2.x Modal
`data-bs-toggle="modal"`

Note: the `-bs-` (bootstrap) identifier in between.

**Reply by bubblypatrick** [reply]
Thank you for the reply. I'm not sure this is addressing my concern as I only have a couple classes like `form-control`, `row`, or `col-6` that don't seem to require the `-bs-` in the documentation. I wasn't aware of `data-bs-toggle` in the docs, and that is good to know so thank you for the info.

My issue is that the web component I have created looks good as is, and is only using the bootstrap 5 css. If someones includes a script to the web component it looks fine. However, if they have bootstrap 3 or 4 as a script reference on their page then my web component will adopt the older styles of bootstrap.

I initially sought to solve this issue with the shadow dom since it can encapsulate styling, but it doesn't seem to be working by default. However, if I manually define some styles within the shadow DOM
` <style>
 .form-control { margin: 30px }
</style>
`
it will behave exactly as expected.

**Comment by dotnetshadow** [comment]
@bubblypatrick Hi there, did you ever find a solution for this?

I'm wondering if namespacing different versions (find and replace) is the only viable solution?
